src/utilities/createRouterRoute.ts

```diff
@@ -1,6 +1,32 @@
 import { readonly } from 'vue'
-import { ResolvedRoute, RouterRoute } from '@/types'
+import { ResolvedRoute } from '@/types/resolved'
+import { RouteUpdate, RouteUpdateOptions, RouterUpdate } from '@/types/routerUpdate'
 
-export function createRouterRoute(route: ResolvedRoute): RouterRoute {
-  return readonly(route)
+export type RouterRoute<TRoute extends ResolvedRoute = ResolvedRoute> = Omit<ResolvedRoute, 'params'> & Readonly<{
+  params: { -readonly [P in keyof TRoute['params']]: TRoute['params'][P] },
+  update: RouteUpdate<TRoute>,
+}>
+
+export function createRouterRoute<TRoute extends ResolvedRoute>(route: TRoute, update: RouterUpdate): RouterRoute<TRoute> {
+  function routeUpdate(keyOrParams: PropertyKey | Partial<ResolvedRoute['params']>, valueOrOptions: any, maybeOptions?: RouteUpdateOptions): Promise<void> {
+    if (typeof keyOrParams === 'object') {
+      return update(route, keyOrParams, valueOrOptions)
+    }
+
+    const updatedParams = readonly({
+      ...route.params,
+      [keyOrParams]: valueOrOptions,
+    })
+    return update(route, updatedParams, maybeOptions)
+  }
+
+  return new Proxy(route as RouterRoute<TRoute>, {
```

Comment:
I think Proxy takes a generic. Does this work?
```suggestion
  return new Proxy<RouterRoute<TRoute>>(route, {
```

Reply:
I tried that but similar to reduce, the starting value is expected to match the generic, which isn't true here

<img width="1010" alt="image" src="https://github.com/kitbagjs/router/assets/6098901/6753646f-7ed8-48c9-a109-87b35ab4474b">